packages/babel-types/src/definitions/typescript.js

```diff
@@ -0,0 +1,416 @@
+// @flow
+
+import defineType, {
+  assertEach,
+  assertNodeType,
+  assertOneOf,
+  assertValueType,
+  chain,
+} from "./index";
+import { functionDeclarationCommon } from "./core";
+import { classMethodOrDeclareMethodCommon } from "./es2015";
+
+const bool = assertValueType("boolean");
+
+function validate(validate) {
+  return { validate };
+}
+
+function typeIs(typeName) {
+  return typeof typeName === "string" ? assertNodeType(typeName) : assertNodeType(...typeName);
+}
+
+function validateType(name) {
+  return validate(typeIs(name));
+}
+
+function validateOptional(validate) {
+  return { validate, optional: true };
+}
+
+function validateOptionalType(typeName) {
+  return { validate: typeIs(typeName), optional: true };
+}
+
+function arrayOf(elementType) {
+  return chain(
+    assertValueType("array"),
+    assertEach(elementType));
+}
+
+function arrayOfType(nodeTypeName) {
+  return arrayOf(typeIs(nodeTypeName));
+}
+
+function validateArrayOfType(nodeTypeName) {
+  return validate(arrayOfType(nodeTypeName));
+}
+
+defineType("TSParameterProperty", {
+  aliases: ["LVal"], // TODO: This isn't usable in general as an LVal. Should have a "Parameter" alias.
+  visitor: ["parameter"],
+  fields: {
+    accessibility: {
+      validate: assertOneOf("public", "private", "protected"),
+      optional: true,
+    },
+    readonly: {
+      validate: assertValueType("boolean"),
+      optional: true,
+    },
+    parameter: {
+      validate: assertNodeType("Identifier", "AssignmentPattern"),
+    },
+  },
+});
+
+defineType("TSDeclareFunction", {
+  aliases: ["Statement", "Declaration"],
+  visitor: ["id", "typeParameters", "params", "returnType"],
+  fields: functionDeclarationCommon,
+});
+
+defineType("TSDeclareMethod", {
+  visitor: ["decorators", "key", "typeParameters", "params", "returnType"],
+  fields: classMethodOrDeclareMethodCommon,
+});
```

Comment:
Are these for overload declarations, or for things that have the `declare` keyword?